src/lib.rs

```diff
@@ -153,24 +157,39 @@ impl RandomXDataset {
                 dataset_start: start,
                 dataset_count: count,
             };
+            let item_count = match result.count() {
+                Ok(v) => v,
+                Err(_) => return Err(CreationError),
+            };
+            // Mirror the assert checks inside randomx_init_dataset call
+            if !((start < item_count && count <= item_count) || (start + item_count <= count)) {
+                return Err(CreationError);
+            }
             unsafe {
+                //no way to check if this fails, c code does not return anything
                 randomx_init_dataset(result.dataset, cache.cache, start, count);
             }
             Ok(result)
         }
     }
 
-    pub fn count(&self) -> c_ulong {
-        unsafe { randomx_dataset_item_count() }
+    pub fn count(&self) -> Result<u64, CreationError> {
+        match unsafe { randomx_dataset_item_count() } {
+            0 => Err(CreationError),
+            x => Ok(x),
+        }
     }
 
-    pub fn get_data(&self) -> Vec<u8> {
+    pub fn get_data(&self) -> Result<Vec<u8>, CreationError> {
         let memory = unsafe { randomx_get_dataset_memory(self.dataset) };
+        if memory.is_null() {
+            return Err(CreationError);
+        }
         let mut result = Vec::new();
         for i in self.dataset_start..self.dataset_count {
             result.push(unsafe { memory.offset(i as isize) } as u8);
         }
```

Comment:
I think safe code should check that the following contracts are honoured:

`dataset_start` and `dataset_count` don't ever let us access out of bounds memory.

We could also probably >> performance by moving this into one unsafe block; allocating the correct amount of memory at once (`Vec::new()` will lead to many re-allocations and unneeded copies) - punted to #9

Reply:
dataset_start and dataset_count is checked when self is/was created by safe code. At this point it should be safe. 
If memory comes back as not null it should also be safe

Reply:
Assigning the values in safe code doesn't make this safe; you have to provide the guarantees yourself as per the Rustinomicon. That said, I looked again, and the bounds are checked in line #159, so I think we're good.